server/src/main/java/org/elasticsearch/action/support/SubscribableListener.java

```diff
@@ -35,6 +35,60 @@
  * completed exceptionally then the exception is passed to subscribed listeners without modification.
+ * <p>
+ * Often this will be used to chain together a sequence of async actions such as in the following example:
+ * <pre>
+ * private void exampleAsyncMethod(String request, List&lt;Long> items, ActionListener&lt;Boolean> finalListener) {
+ *     SubscribableListener
+ *
+ *         // Start the chain and run the first step by creating a SubscribableListener using newForked():
+ *         .&lt;String>newForked(l -> firstAsyncStep(request, l))
+ *
+ *         // Run a second step when the first step completes using andThen(); if the first step fails then the exception falls through to
+ *         // the end without executing the intervening steps.
+ *         .&lt;Integer>andThen((l, firstStepResult) -> secondAsyncStep(request, firstStepResult, l))
+ *
+ *         // Run another step when the second step completes with another andThen() call; as above this only runs if the first two steps
+ *         // succeed.
+ *         .&lt;Boolean>andThen((l, secondStepResult) -> {
+ *             if (condition) {
+ *                 // Steps are exception-safe: an exception thrown here will be passed to the listener rather than escaping to the
+ *                 // caller.
+ *                 throw new IOException("failure");
+ *             }
+ *
+ *             // Steps can fan out to multiple subsidiary async actions using utilities like RefCountingListener.
+ *             final var result = new AtomicBoolean();
+ *             try (var listeners = new RefCountingListener(l.map(v -> result.get()))) {
+ *                 for (final var item : items) {
+ *                     thirdAsyncStep(secondStepResult, item, listeners.acquire());
+ *                 }
+ *             }
+ *         })
+ *
+ *         // Synchronous (non-forking) steps which do not return a result can be expressed using andThenAccept() with a consumer:
+ *         .andThenAccept(thirdStepResult -> {
+ *             if (condition) {
+ *                 // andThenAccept() is also exception-safe
+ *                 throw new ElasticsearchException("some other problem");
+ *             }
+ *             consumeThirdStepResult(thirdStepResult);
+ *         })
+ *
+ *         // Synchronous (non-forking) steps which do return a result can be expressed using andThenApply() with a function:
+ *         .andThenApply(voidFromStep4 -> {
+ *             if (condition) {
+ *                 // andThenApply() is also exception-safe
+ *                 throw new IllegalArgumentException("failure");
+ *             }
+ *             return computeFifthStepResult();
+ *         })
+ *
+ *         // To complete the chain, add the outer listener which will be completed with the result of the previous step if all steps were
+ *         // successful, or the exception if any step failed.
+ *         .addListener(finalListener);
+ * }
+ * </pre>
```

Comment:
The class comment addition is certainly nice, but I jumped to the method comments when I was trying to figure out what was happening -- so `newForked` and `andThen`.

`andThen`'s documentation is pretty confusing -- it's very outdated, and looks like the @code doesn't really work? I'm not too familiar with how to reference things better. I know @link references must be updated if the referenced code changes, but @code doesn't seem to have that feature.

The surprised I had was that I thought SubscribableListener's purpose was to register Listeners and eventually notify them all at once when an event happens. But this use case is completely different. So SubscribableListener is handling two different concepts at once, which is very unexpected.

Reply:
Huh, I misread `andThen`'s comment. Quite a lot. I was looking for a method called `forkAction1` -- thinking it was `newForked` -- and looking for type declarations `L` and `R` 😌  That's what I get for skimming, I guess, and not using the pretty view.

> In practice if you want to write a sequence of async steps then it's best to create a listener for each step and then subscribe to it in the next step. The only other options are to write the steps in https://github.com/elastic/elasticsearch/pull/99623 which is just horrible, or to nest each step within the previous which is also fairly awful to read. See also the JDK's own CompletionStage, which is implementing roughly the same pattern (we don't use it directly because it catches Error which creates awful bugs).

Thanks for the explanation